tests/test_record.py

```diff
@@ -155,36 +252,154 @@ def test(self):
+
+    def test_spider_attributes_recursive(self):
+        # Recursive calls including private variables
+        with CaseSpider() as spider:
+            spider.start_requests("""
+                self.__page = 0
+                self.param = 0
+                self._base_url = 'www.nothing.com'
+                yield scrapy.Request('data:text/plain,', callback=self.parse)
+            """)
+            spider.parse("""
+                self.param += 1
+                reqs = self.parse_item()
+                for r in reqs:
+                    yield r
+            """)
+            spider.parse_item("""
+                self.__page += 1
+                if self.__page > 3:
+                    self.end = True
+                    yield {'a': 4}
+                    return
+                for i in range(3):
+                    yield {'b': '%s_%s;'%(self.__page, i)}
+                yield scrapy.Request('data:,%s;'%(self.__page),
+                                      callback=self.parse)
+                                         """)
+            spider.record()
+            spider.test()
+
+        # Recursive calls including private variables using getattr
+        with CaseSpider() as spider:
+            spider.start_requests("""
+                self.param = 0
+                self._base_url = 'www.nothing.com'
+                yield scrapy.Request('data:text/plain,', callback=self.parse)
+            """)
+            spider.parse("""
+                self.param += 1
+                reqs = self.parse_item()
+                for r in reqs:
+                    yield r
+            """)
+            spider.parse_item("""
+                self.__page = getattr(self, '_MySpider__page', 0) + 1
+                if self.__page > 3:
+                    self.end = True
+                    yield {'a': 4}
+                    return
+                for i in range(3):
+                    yield {'b': '%s_%s;'%(self.__page, i)}
+                yield scrapy.Request('data:,%s;'%(self.__page),
+                                      callback=self.parse)
+                                         """)
+            spider.record()
+            spider.test()
+
+        # Testing spider with rescursive spiders
+        with CaseSpider() as spider:
+            spider._write_generic_spider("""
+import scrapy
+
+
+class HijSpider(scrapy.Spider):
+    name = 'hij'
+
+    custom_settings = dict(
+        SPIDER_MIDDLEWARES={
+            'myproject.middleware.AutounitMiddleware': 950,
+        }
+    )
+
+    def __init__(self, *pargs, **kwargs):
+        super(HijSpider, self).__init__(*pargs, **kwargs) # py 2.7 compatible
+        self.i = 0
+
+    def start_requests(self):
+        yield self.next_req()
+
+    def next_req(self):
+        self.i += 1
+        return scrapy.Request('data:text/plain,hi', dont_filter=True)
+
+    def parse(self, response):
+        if self.i < 3:
+            yield self.next_req()
+
+                """)
+
```

Comment:
@mabelvj, many thanks for keeping up the work on this! 💪 
I think the first four cases don't actually test the spider attribute issue because they would work with the current version of autounit.
Maybe a case that tests this issue would be something like this:
```
import scrapy


class AttrsSpider(scrapy.Spider):
    name = 'attrs'

    def __init__(self, *pargs, **kwargs):
        self.page_number = 0

    def start_requests(self):
        yield scrapy.Request('data:text/plain,')

    def parse(self, response):
        self.page_number += 1
        yield {
            'page_number': self.page_number
        }
        if self.page_number < 3:
            yield scrapy.Request('data:text/plain,', dont_filter=True)
```
What do you think about creating a single test case reflecting this example, leaving the generic spider aggregate out of this PR's scope?

Reply:
The version that is actually tested is the one in the files, not the installed. `middleware.py`  is copied in `_write_mw` in #L109, so we can run the test on the actual files we have in the branch without needing to install it each time.  I have inspected the internal results and they are correct. You can inspect the output setting the verbosity in `record` and `test`.

It might be out of the scope of the PR, but I developed that code to make easier the debugging and thought it might help. I can put it on a separate PR if not. What do you think?